### DISCUSSION TITLE
Missing Columns on Custom Formatter

### DISCUSSION THREAD

**Discussion by jfvoliveira** [opcodesio/log-viewer#305] [Q&A] [answered]
Hey there. I am trying to create a custom log formatter but I am hitting a wall that I cannot overcome and I need help.

Regex link: [https://regexr.com/7omeu](https://regexr.com/7omeu)

This is my custom log:

```php
class LogViewerCustom extends Log
{
    public static string $name = 'Custom';

    public static string $regex = '/^\[(?P<datetime>[\d\-+ :]+)\] \[(?P<env>\w+)\.(?P<level>\w+)\] \[(?P<code>.{5})\] (?P<message>.+)\s(?P<payload>\{.*\})?$/';

    /** @var array|string[][] */
    public static array $columns = [
        ['label' => 'Level', 'data_path' => 'level'],
        ['label' => 'Date/Time', 'data_path' => 'datetime'],
        ['label' => 'Env', 'data_path' => 'env'],
        ['label' => 'Code', 'data_path' => 'code'],
        ['label' => 'Message', 'data_path' => 'message'],
        ['label' => 'Payload', 'data_path' => 'payload'],
    ];
}
```
and this is the laravel.log:
```log
[2023-12-07 04:24:02] [local.ERROR] [12345] error {"message":"oops"}
[2023-12-07 04:21:59] [local.INFO] [ABCDE] some info {"info":"some"}
[2023-12-07 04:24:02] [local.INFO] [ABCDE] here be dragons {"name":"Smaug"}
[2023-12-07 04:24:02] [local.DEBUG] [DBG01] no payload  
[2023-12-07 04:24:02] [local.DEBUG] [DBG01] payload {"foo":"bar"}
```
This is the result:

![logs](https://github.com/opcodesio/log-viewer/assets/3233802/941968b6-bafb-4326-bedf-4a39f2af1178)

No data on Env nor Code. And the Payload column is missing. But the log row is read correctly:

![logs2](https://github.com/opcodesio/log-viewer/assets/3233802/d9e97f83-f91a-4806-b314-8963a3aff727)

But if I do a dd on the fillMatches method
```php
protected function fillMatches(array $matches = []): void
{
    parent::fillMatches($matches);
    
    dd($matches);
}
```
I get (on the Dev Tools response):
```php
array:11 [ // app\Logging\LogViewerCustom.php:27
  0 => "[2023-12-07 04:24:02] [local.DEBUG] [DBG02] payload {"foo":"bar"}"
  "datetime" => "2023-12-07 04:24:02"
  1 => "2023-12-07 04:24:02"
  "env" => "local"
  2 => "local"
  "level" => "DEBUG"
  3 => "DEBUG"
  "code" => "DBG02"
  4 => "DBG02"
  "message" => "payload {"foo":"bar"}"
  5 => "payload {"foo":"bar"}"
]
```
So shouldn't the data_path for env and code be there?

What am I missing?

Thanks!

**Comment by arukompas** [comment, marked as answer]
Hey @jfvoliveira , glad you're experimenting with custom logs!

So, I'm not sure if I missed it in the docs or not, but it's not enough to just provide extra regex named groups. You should also update the `fillMatches()` method that you mentioned above, to something like this:

```php
protected function fillMatches(array $matches = []): void
{
    // the parent method only parses datetime, level, and message
    parent::fillMatches($matches);
    
    // for other properties, fill them in $this->context or $this->extra
    $this->context = [
        'env' => $matches['env'] ?? '',
        'code' => $matches['code'] ?? '',
        'payload' => json_decode($matches['payload'] ?? 'null'),
    ];
    
    // DO NOT assign properties outside of 'context' or 'extra' arrays, as it will not get returned to the frontend.
    $this->env = $matches['env'];  // Do not do this
}
```

If you have any other custom properties/matches to define on the log, do it on the `$this->context` or `$this->extra` array. The `$this->extra` is usually used by Log Viewer, but you're free to fill it as well.

So, now that you have the properties clearly defined inside the context or extra arrays, it's time to update the column definitions:

```php
    /** @var array|string[][] */
    public static array $columns = [
        ['label' => 'Level', 'data_path' => 'level'],
        ['label' => 'Date/Time', 'data_path' => 'datetime'],
        ['label' => 'Env', 'data_path' => 'context.env'],
        ['label' => 'Code', 'data_path' => 'context.code'],
        ['label' => 'Message', 'data_path' => 'message'],
        ['label' => 'Payload', 'data_path' => 'context.payload'],
    ];
```

Let me know if this helps, or if you have any more questions :)

**Reply by jfvoliveira** [reply]
Hey @arukompas . Thank you for your response.
I read the docs but I thought that those `context.<field>` were only required if the `<field>` was not part of the named groups of the regex (in my case they were).
Already got this working. Many thanks!

**Reply by arukompas** [reply]
I'll update the docs to be more clear, thanks for letting me know! 🙌

**Reply by jfvoliveira** [reply]
No problem. I'm the one thanking you.

Btw, just two more things I saw:

![logs3](https://github.com/opcodesio/log-viewer/assets/3233802/ca7af703-851f-4020-b09e-dab8cae75890)

The Payload column is pushed to the right, and the title is hidden behind the sort/per page options and the details row is 2 colspan short The image says 1 colspan but in this case the main row has 8 columns and the details row is `colspan="6"`.

Thanks for your help and keep up the good work!

**Reply by arukompas** [reply]
hey @jfvoliveira , please try upgrading to **v3.1.10** (latest), it should fix the layout issues mentioned above :)

**Reply by jfvoliveira** [reply]
Yeah, can confirm. It's fixed now. Many thanks! 👍🏻
